// @connectlive/connectlive-web-sdk/es6/modules/local/local-media.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { ClientError } from '../../utils/error-exception';
import { logger, Loggable } from '../../utils/logger';
import { LocalAudio } from './local-audio';
import { LocalVideo } from './local-video';
class LocalMedia {
    constructor(constraints = {}) {
        this.room = null;
        this.audio = undefined;
        this.video = undefined;
        if (constraints.video) {
            if (constraints.video === true)
                constraints.video = {};
        }
        if (constraints.audio) {
            if (constraints.audio === true)
                constraints.audio = {};
        }
        this.constraints = constraints;
    }
    getDevices() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield navigator.mediaDevices.enumerateDevices();
        });
    }
    getMicDevices() {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.getDevices()).filter(device => device.kind === 'audioinput' && device.deviceId);
        });
    }
    getSpeakerDevices() {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.getDevices()).filter(device => device.kind === 'audiooutput' && device.deviceId);
        });
    }
    getCameraDevices() {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.getDevices()).filter(device => device.kind === 'videoinput' && device.deviceId);
        });
    }
    switchCamera(deviceId) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.constraints) {
                if (this.constraints.video) {
                    this.constraints.video = Object.assign(this.constraints.video, { deviceId: { exact: deviceId } });
                    if (this.room && this.video) {
                        this.video.stop();
                        yield this.room.unpublish([this.video]);
                    }
                    let mediaStream;
                    try {
                        mediaStream = yield navigator.mediaDevices.getUserMedia({
                            video: this.constraints.video,
                            audio: false
                        });
                    }
                    catch (err) {
                        const e = err;
                        const message = e.message;
                        logger.error('Failed to switchCamera. ', e.name, message);
                        throw new ClientError(1203, `getUserMedia 실패(${message})`, e);
                    }
                    const [videoTrack] = mediaStream.getVideoTracks();
                    if (videoTrack) {
                        if (videoTrack.readyState === 'ended') {
                            throw new ClientError(1206, '비디오 트랙을 획득했지만 종료 상태입니다.');
                        }
                        this.video = new LocalVideo(videoTrack, 'multiple');
                        logger.debug('switch video track', 'id=' + videoTrack.id, 'label=' + videoTrack.label, 'enabled=' + videoTrack.enabled, 'kind=' + videoTrack.kind, 'muted=' + videoTrack.muted, 'readyState=' + videoTrack.readyState);
                        if (this.room)
                            yield this.room.publish([this.video]);
                    }
                }
            }
        });
    }
    switchMic(deviceId) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.constraints) {
                if (this.constraints.audio) {
                    this.constraints.audio = Object.assign(this.constraints.audio, { deviceId: { exact: deviceId } });
                    if (this.room && this.audio) {
                        this.audio.stop();
                        yield this.room.unpublish([this.audio]);
                    }
                    let mediaStream;
                    try {
                        mediaStream = yield navigator.mediaDevices.getUserMedia({
                            video: false,
                            audio: this.constraints.audio
                        });
                    }
                    catch (err) {
                        const e = err;
                        const message = e.message;
                        logger.error('Failed to switchMic. ', e.name, message);
                        throw new ClientError(1203, `getUserMedia 실패(${message})`, e);
                    }
                    const [audioTrack] = mediaStream.getAudioTracks();
                    if (audioTrack) {
                        if (audioTrack.readyState === 'ended') {
                            throw new ClientError(1205, '오디오 트랙을 획득했지만 종료 상태입니다.');
                        }
                        this.audio = new LocalAudio(audioTrack, 'single');
                        logger.debug('switch audio track', 'id=' + audioTrack.id, 'label=' + audioTrack.label, 'enabled=' + audioTrack.enabled, 'kind=' + audioTrack.kind, 'muted=' + audioTrack.muted, 'readyState=' + audioTrack.readyState);
                        if (this.room)
                            yield this.room.publish([this.audio]);
                    }
                }
            }
        });
    }
    _getUserMedia() {
        return __awaiter(this, void 0, void 0, function* () {
            this.stop();
            try {
                const mediaStream = yield navigator.mediaDevices.getUserMedia(this.constraints);
                const [audioTrack] = mediaStream.getAudioTracks();
                const [videoTrack] = mediaStream.getVideoTracks();
                logger.debug('got mediaStream', 'id=' + mediaStream.id, 'active=' + mediaStream.active);
                if (audioTrack) {
                    logger.debug('got audio track', 'id=' + audioTrack.id, 'label=' + audioTrack.label, 'enabled=' + audioTrack.enabled, 'kind=' + audioTrack.kind, 'muted=' + audioTrack.muted, 'readyState=' + audioTrack.readyState);
                    if (audioTrack.readyState === 'ended') {
                        throw new ClientError(1205, '오디오 트랙을 획득했지만 종료 상태입니다.');
                    }
                }
                if (videoTrack) {
                    logger.debug('got video track', 'id=' + videoTrack.id, 'label=' + videoTrack.label, 'enabled=' + videoTrack.enabled, 'kind=' + videoTrack.kind, 'muted=' + videoTrack.muted, 'readyState=' + videoTrack.readyState);
                    if (videoTrack.readyState === 'ended') {
                        throw new ClientError(1206, '비디오 트랙을 획득했지만 종료 상태입니다.');
                    }
                }
                this.audio = audioTrack ? new LocalAudio(audioTrack, 'single') : undefined;
                this.video = videoTrack ? new LocalVideo(videoTrack, 'multiple') : undefined;
            }
            catch (err) {
                const e = err;
                const message = e.message;
                logger.error('Failed to getUserMedia. ', e.name, message);
                throw new ClientError(1203, `getUserMedia 실패(${message})`, e);
            }
        });
    }
    stop() {
        var _a, _b;
        (_a = this.audio) === null || _a === void 0 ? void 0 : _a.stop();
        (_b = this.video) === null || _b === void 0 ? void 0 : _b.stop();
    }
    _setRoom(room) {
        this.room = room;
    }
    setMediaStream(mediaStream) {
        const [audioTrack] = mediaStream.getAudioTracks();
        const [videoTrack] = mediaStream.getVideoTracks();
        this.audio = audioTrack ? new LocalAudio(audioTrack, 'single') : undefined;
        this.video = videoTrack ? new LocalVideo(videoTrack, 'multiple') : undefined;
        if (audioTrack)
            logger.debug('set media stream track', 'id=' + audioTrack.id, 'label=' + audioTrack.label, 'enabled=' + audioTrack.enabled, 'kind=' + audioTrack.kind, 'muted=' + audioTrack.muted, 'readyState=' + audioTrack.readyState);
        if (videoTrack)
            logger.debug('set media stream track', 'id=' + videoTrack.id, 'label=' + videoTrack.label, 'enabled=' + videoTrack.enabled, 'kind=' + videoTrack.kind, 'muted=' + videoTrack.muted, 'readyState=' + videoTrack.readyState);
    }
}
__decorate([
    Loggable('debug')
], LocalMedia.prototype, "getDevices", null);
__decorate([
    Loggable('debug')
], LocalMedia.prototype, "getMicDevices", null);
__decorate([
    Loggable('debug')
], LocalMedia.prototype, "getSpeakerDevices", null);
__decorate([
    Loggable('debug')
], LocalMedia.prototype, "getCameraDevices", null);
__decorate([
    Loggable('info')
], LocalMedia.prototype, "switchCamera", null);
__decorate([
    Loggable('info')
], LocalMedia.prototype, "switchMic", null);
__decorate([
    Loggable('info')
], LocalMedia.prototype, "_getUserMedia", null);
__decorate([
    Loggable('info')
], LocalMedia.prototype, "stop", null);
__decorate([
    Loggable('info')
], LocalMedia.prototype, "setMediaStream", null);
export { LocalMedia };
