// @connectlive/connectlive-web-sdk/es6/utils/room-event-enum.js
export var RoomEventName;
(function (RoomEventName) {
    RoomEventName["CONNECTING"] = "connecting";
    RoomEventName["CONNECTED"] = "connected";
    RoomEventName["DISCONNECTED"] = "disconnected";
    RoomEventName["PARTICIPANTENTERED"] = "participantEntered";
    RoomEventName["PARTICIPANTLEFT"] = "participantLeft";
    RoomEventName["LOCALVIDEOPUBLISHED"] = "localVideoPublished";
    RoomEventName["LOCALVIDEOUNPUBLISHED"] = "localVideoUnpublished";
    RoomEventName["LOCALAUDIOPUBLISHED"] = "localAudioPublished";
    RoomEventName["LOCALAUDIOUNPUBLISHED"] = "localAudioUnpublished";
    RoomEventName["REMOTEVIDEOPUBLISHED"] = "remoteVideoPublished";
    RoomEventName["REMOTEVIDEOUNPUBLISHED"] = "remoteVideoUnpublished";
    RoomEventName["REMOTEAUDIOPUBLISHED"] = "remoteAudioPublished";
    RoomEventName["REMOTEAUDIOUNPUBLISHED"] = "remoteAudioUnpublished";
    RoomEventName["REMOTEAUDIOLEVELUPDATED"] = "remoteAudioLevelUpdated";
    RoomEventName["REMOTEAUDIOSUBSCRIBED"] = "remoteAudioSubscribed";
    RoomEventName["REMOTEAUDIOUNSUBSCRIBED"] = "remoteAudioUnsubscribed";
    RoomEventName["REMOTEVIDEOSTATECHANGED"] = "remoteVideoStateChanged";
    RoomEventName["REMOTEAUDIOSTATECHANGED"] = "remoteAudioStateChanged";
    RoomEventName["ERROR"] = "error";
    RoomEventName["RECORDFAILED"] = "recordFailed";
    RoomEventName["USERMESSAGE"] = "userMessage";
    RoomEventName["STAT"] = "stat";
})(RoomEventName || (RoomEventName = {}));
;
